File: frontend/src/types/roleModelConfig.ts
/**
 * 角色模型配置类型定义
 */

// ==================== 基础类型 ====================

/** 角色类型 */
export type RoleType = 'analyst' | 'debater' | 'decision_maker'

/** 模型配置引用 */
export interface ModelConfigRef {
  provider: string
  model_name: string
  temperature?: number
  max_tokens?: number
  timeout?: number
}

/** 角色模型配置引用（用于预设） */
export interface RoleModelConfigRef {
  provider: string
  model_name: string
  temperature?: number
  max_tokens?: number
  custom_prompt?: string
}

// ==================== 角色配置 ====================

/** 角色信息 */
export interface RoleInfo {
  id: string
  name: string
  type: RoleType
  icon: string
  description: string
}

/** 角色模型配置 */
export interface RoleModelConfig {
  id?: string
  role_id: string
  role_type: RoleType
  role_name: string
  provider: string
  model_name: string
  model_display_name?: string
  temperature: number
  max_tokens: number
  timeout: number
  is_enabled: boolean
  is_default: boolean
  description?: string
  created_at?: string
  updated_at?: string
}

/** 角色配置响应（包含角色信息和当前配置） */
export interface RoleConfigResponse {
  role_id: string
  role_name: string
  role_type: RoleType
  icon: string
  description: string
  current_config: RoleModelConfig | null
}

/** 所有角色配置响应 */
export interface RoleConfigsResponse {
  analysts: RoleConfigResponse[]
  debaters: RoleConfigResponse[]
  decision_makers: RoleConfigResponse[]
}

/** 角色配置更新请求 */
export interface RoleModelConfigUpdate {
  provider: string
  model_name: string
  model_display_name?: string
  temperature: number
  max_tokens: number
  timeout: number
  is_enabled: boolean
  description?: string
}

// ==================== 预设 ====================

/** 模型选择预设 */
export interface ModelSelectionPreset {
  id?: string
  name: string
  description?: string
  preset_type: 'system' | 'user'
  role_configs: Record<string, RoleModelConfigRef>
  default_provider: string
  default_model: string
  is_active: boolean
  is_system: boolean
  created_at?: string
  updated_at?: string
}

/** 创建预设请求 */
export interface ModelSelectionPresetCreate {
  name: string
  description?: string
  role_configs: Record<string, RoleModelConfigRef>
  default_provider: string
  default_model: string
}

/** 更新预设请求 */
export interface ModelSelectionPresetUpdate {
  name?: string
  description?: string
  role_configs?: Record<string, RoleModelConfigRef>
  default_provider?: string
  default_model?: string
  is_active?: boolean
}

/** 应用预设响应 */
export interface ApplyPresetResponse {
  success: boolean
  message: string
  applied_configs: number
}

// ==================== 运行时配置 ====================

/** 当前生效的配置 */
export interface ActiveModelConfig {
  preset_id?: string
  preset_name?: string
  configs: {
    analysts: Record<string, ModelConfigRef>
    debaters: Record<string, ModelConfigRef>
    decision_makers: Record<string, ModelConfigRef>
  }
  default_model?: ModelConfigRef
}

// ==================== 预定义角色 ====================

/** 分析师角色 */
export const ANALYST_ROLES: RoleInfo[] = [
  { id: 'market', name: '技术面分析师', type: 'analyst', icon: 'TrendCharts', description: '技术分析、价格走势分析' },
  { id: 'fundamentals', name: '基本面分析师', type: 'analyst', icon: 'Document', description: '财务报表、基本面分析' },
  { id: 'capital_flow', name: '资金流分析师', type: 'analyst', icon: 'Money', description: '主力资金、北向资金分析' },
  { id: 'news', name: '新闻分析师', type: 'analyst', icon: 'Message', description: '新闻舆情分析' },
  { id: 'social', name: '社媒分析师', type: 'analyst', icon: 'ChatDotRound', description: '社交媒体情绪分析' },
  { id: 'market_trend', name: '大盘走势分析师', type: 'analyst', icon: 'DataLine', description: '大盘指数、行业分析' }
]

/** 辩论者角色 */
export const DEBATER_ROLES: RoleInfo[] = [
  { id: 'bull', name: '多头分析师', type: 'debater', icon: 'Top', description: '看涨观点分析' },
  { id: 'bear', name: '空头分析师', type: 'debater', icon: 'Bottom', description: '看跌观点分析' }
]

/** 决策者角色 */
export const DECISION_ROLES: RoleInfo[] = [
  { id: 'investment_manager', name: '投资经理', type: 'decision_maker', icon: 'User', description: '投资决策' },
  { id: 'portfolio_manager', name: '组合经理', type: 'decision_maker', icon: 'Collection', description: '组合管理' },
  { id: 'risk_manager', name: '风险经理', type: 'decision_maker', icon: 'Warning', description: '风险评估' }
]

/** 所有角色 */
export const ALL_ROLES: RoleInfo[] = [...ANALYST_ROLES, ...DEBATER_ROLES, ...DECISION_ROLES]

/** 角色类型显示名称 */
export const ROLE_TYPE_NAMES: Record<RoleType, string> = {
  analyst: '分析师',
  debater: '辩论者',
  decision_maker: '决策者'
}

/** 获取角色信息 */
export function getRoleInfo(roleId: string): RoleInfo | undefined {
  return ALL_ROLES.find(role => role.id === roleId)
}

/** 获取角色类型 */
export function getRoleType(roleId: string): RoleType | undefined {
  return getRoleInfo(roleId)?.type
}

File: app/models/role_model_config.py
"""
角色模型配置数据模型
"""

from datetime import datetime
from typing import Optional, Dict, Any, List
from pydantic import BaseModel, Field, ConfigDict
from bson import ObjectId
from app.utils.timezone import now_tz
from .user import PyObjectId


class ModelConfigRef(BaseModel):
    """模型配置引用"""
    provider: str = Field(..., description="模型厂家")
    model_name: str = Field(..., description="模型名称")
    temperature: Optional[float] = Field(default=0.7, description="温度参数")
    max_tokens: Optional[int] = Field(default=4000, description="最大token数")
    timeout: Optional[int] = Field(default=180, description="超时时间(秒)")
    
    model_config = ConfigDict(populate_by_name=True)


class RoleModelConfigRef(BaseModel):
    """角色模型配置引用（用于预设）"""
    provider: str = Field(..., description="模型厂家")
    model_name: str = Field(..., description="模型名称")
    temperature: Optional[float] = Field(default=0.7, description="温度参数")
    max_tokens: Optional[int] = Field(default=4000, description="最大token数")
    custom_prompt: Optional[str] = Field(None, description="角色特定的系统提示词覆盖")
    
    model_config = ConfigDict(populate_by_name=True)


class RoleModelConfig(BaseModel):
    """角色模型配置"""
    id: Optional[PyObjectId] = Field(default_factory=PyObjectId, alias="_id")
    
    # 角色标识
    role_id: str = Field(..., description="角色ID，如：market, bull, investment_manager")
    role_type: str = Field(..., description="角色类型：analyst/debater/decision_maker")
    role_name: str = Field(..., description="角色显示名称")
    
    # 模型配置
    provider: str = Field(..., description="模型厂家")
    model_name: str = Field(..., description="模型名称")
    model_display_name: Optional[str] = Field(None, description="模型显示名称")
    
    # 模型参数
    temperature: float = Field(default=0.7, ge=0.0, le=2.0, description="温度参数")
    max_tokens: int = Field(default=4000, description="最大token数")
    timeout: int = Field(default=180, description="超时时间(秒)")
    
    # 启用状态
    is_enabled: bool = Field(default=True, description="是否启用")
    is_default: bool = Field(default=False, description="是否为该类型角色的默认配置")
    
    # 元数据
    description: Optional[str] = Field(None, description="配置描述")
    created_at: datetime = Field(default_factory=now_tz)
    updated_at: datetime = Field(default_factory=now_tz)
    created_by: Optional[str] = Field(None, description="创建者")
    
    model_config = ConfigDict(populate_by_name=True, arbitrary_types_allowed=True)


class ModelSelectionPreset(BaseModel):
    """模型选择预设方案"""
    id: Optional[PyObjectId] = Field(default_factory=PyObjectId, alias="_id")
    
    name: str = Field(..., description="预设名称")
    description: Optional[str] = Field(None, description="预设描述")
    
    # 预设类型：system（系统预设）/ user（用户自定义）
    preset_type: str = Field(default="user", description="预设类型")
    
    # 角色配置映射
    role_configs: Dict[str, RoleModelConfigRef] = Field(default_factory=dict, description="角色配置映射")
    # 格式：{"market": {"provider": "openai", "model_name": "gpt-4"}, ...}
    
    # 默认配置（当角色没有特定配置时使用）
    default_provider: str = Field(..., description="默认厂家")
    default_model: str = Field(..., description="默认模型")
    
    is_active: bool = Field(default=True, description="是否激活")
    is_system: bool = Field(default=False, description="是否为系统预设")
    
    created_at: datetime = Field(default_factory=now_tz)
    updated_at: datetime = Field(default_factory=now_tz)
    created_by: Optional[str] = Field(None, description="创建者")
    
    model_config = ConfigDict(populate_by_name=True, arbitrary_types_allowed=True)


class AnalysisModelConfig(BaseModel):
    """分析任务模型配置（运行时配置）"""
    id: Optional[PyObjectId] = Field(default_factory=PyObjectId, alias="_id")
    
    # 关联的分析任务或用户
    task_id: Optional[str] = Field(None, description="任务ID")
    user_id: Optional[str] = Field(None, description="用户ID")
    
    # 使用的预设
    preset_id: Optional[str] = Field(None, description="预设ID")
    
    # 实际使用的模型配置（展开后的完整配置）
    analyst_models: Dict[str, ModelConfigRef] = Field(default_factory=dict, description="分析师模型配置")
    debater_models: Dict[str, ModelConfigRef] = Field(default_factory=dict, description="辩论者模型配置")
    decision_models: Dict[str, ModelConfigRef] = Field(default_factory=dict, description="决策者模型配置")
    
    # 默认模型（当特定角色没有配置时使用）
    default_model: Optional[ModelConfigRef] = Field(None, description="默认模型配置")
    
    created_at: datetime = Field(default_factory=now_tz)
    updated_at: datetime = Field(default_factory=now_tz)
    
    model_config = ConfigDict(populate_by_name=True, arbitrary_types_allowed=True)


# 请求/响应模型

class RoleModelConfigUpdate(BaseModel):
    """角色模型配置更新请求"""
    provider: str = Field(..., description="模型厂家")
    model_name: str = Field(..., description="模型名称")
    model_display_name: Optional[str] = Field(None, description="模型显示名称")
    temperature: float = Field(default=0.7, ge=0.0, le=2.0)
    max_tokens: int = Field(default=4000)
    timeout: int = Field(default=180)
    is_enabled: bool = Field(default=True)
    description: Optional[str] = Field(None)


class RoleModelConfigResponse(BaseModel):
    """角色模型配置响应"""
    id: str
    role_id: str
    role_type: str
    role_name: str
    provider: str
    model_name: str
    model_display_name: Optional[str] = None
    temperature: float
    max_tokens: int
    timeout: int
    is_enabled: bool
    is_default: bool
    description: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class BatchRoleModelConfigUpdate(BaseModel):
    """批量角色模型配置更新请求"""
    configs: List[RoleModelConfigUpdate] = Field(..., description="配置列表")


class ModelSelectionPresetCreate(BaseModel):
    """模型选择预设创建请求"""
    name: str = Field(..., description="预设名称")
    description: Optional[str] = Field(None)
    role_configs: Dict[str, RoleModelConfigRef] = Field(default_factory=dict)
    default_provider: str = Field(...)
    default_model: str = Field(...)


class ModelSelectionPresetUpdate(BaseModel):
    """模型选择预设更新请求"""
    name: Optional[str] = Field(None)
    description: Optional[str] = Field(None)
    role_configs: Optional[Dict[str, RoleModelConfigRef]] = Field(None)
    default_provider: Optional[str] = Field(None)
    default_model: Optional[str] = Field(None)
    is_active: Optional[bool] = Field(None)


class ModelSelectionPresetResponse(BaseModel):
    """模型选择预设响应"""
    id: str
    name: str
    description: Optional[str] = None
    preset_type: str
    role_configs: Dict[str, RoleModelConfigRef]
    default_provider: str
    default_model: str
    is_active: bool
    is_system: bool
    created_at: datetime
    updated_at: datetime


class ActiveModelConfigResponse(BaseModel):
    """当前生效的模型配置响应"""
    preset_id: Optional[str] = None
    preset_name: Optional[str] = None
    configs: Dict[str, Dict[str, Any]] = Field(default_factory=dict)
    default_model: Optional[ModelConfigRef] = None


class ApplyPresetResponse(BaseModel):
    """应用预设响应"""
    success: bool
    message: str
    applied_configs: int
